Refactor marked text paint logic to reduce code duplication

rr-codes opened the pull request:
#### a3ab05dda99ae804dcbc3a5d4d003d42b6c4ec90
<pre>
Refactor marked text paint logic to reduce code duplication
<a href="https://bugs.webkit.org/show_bug.cgi?id=252540">https://bugs.webkit.org/show_bug.cgi?id=252540</a>
rdar://105411314

Reviewed by NOBODY (OOPS!).

This builds upon <a href="https://github.com/WebKit/WebKit/pull/10184">https://github.com/WebKit/WebKit/pull/10184</a> to reduce
code duplication when creating marked text compositions.

* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::fillCompositionUnderline const):
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintCompositionUnderlines):
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintCompositionUnderline):
* Source/WebCore/rendering/TextBoxPainter.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/MarkedTextComposition.h: Added.
* Source/WebKit/UIProcess/Cocoa/MarkedTextComposition.mm: Added.
(WebKit::compositionRangeComparator):
(WebKit::markedTextCompositionModeForString):
(WebKit::extractUnderlines):
(WebKit::compositionHighlights):
(WebKit::extractMarkedTextComposition):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setMarkedText:selectedRange:]):
(shouldExtractMarkedTextComposition):
(-[WKContentView setAttributedMarkedText:selectedRange:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::shouldExtractMarkedTextComposition):
(WebKit::WebViewImpl::setMarkedText):
(WebKit::extractInitialUnderlines): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3ab05dda99ae804dcbc3a5d4d003d42b6c4ec90

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108958 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18038 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41782 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/486 "Reverted pull request changes (failure)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118249 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112841 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19494 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9333 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101189 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114714 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14620 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97851 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42741 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96592 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29492 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84500 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10820 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30841 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11568 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7777 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16956 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50437 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13164 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/Cocoa/MarkedTextComposition.mm:
```diff
@@ -0,0 +1,166 @@
+
+namespace WebKit {
+
+enum class MarkedTextCompositionMode {
```
Minor nit - `: uint8_t`

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -5686,17 +5687,35 @@ - (void)setMarkedTextStyle:(NSDictionary *)styleDictionary
+        composition = compositionHighlights(markedText);
+
+    WTF::switchOn(composition, [&](std::monostate) {
+        [self _setMarkedText:markedText.string underlines:Vector<WebCore::CompositionUnderline> { } highlights:Vector<WebCore::CompositionHighlight> { } selectedRange:selectedRange];
```
Nit - does it work if you replace the `Vector<WebCore::CompositionUnderline> { }` with just `{ }`? (If so, it might make this a bit cleaner)

(Ditto re: `Vector<WebCore::CompositionHighlight> { }`)

rr-codes committed:
Refactor marked text paint logic to reduce code duplication
https://bugs.webkit.org/show_bug.cgi?id=252540
rdar://105411314

Reviewed by NOBODY (OOPS!).

This builds upon https://github.com/WebKit/WebKit/pull/10184 to reduce
code duplication when creating marked text compositions.

* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter<TextBoxPath>::fillCompositionUnderline const):
(WebCore::TextBoxPainter<TextBoxPath>::paintCompositionUnderlines):
(WebCore::TextBoxPainter<TextBoxPath>::paintCompositionUnderline):
* Source/WebCore/rendering/TextBoxPainter.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/MarkedTextComposition.h: Added.
* Source/WebKit/UIProcess/Cocoa/MarkedTextComposition.mm: Added.
(WebKit::compositionRangeComparator):
(WebKit::markedTextCompositionModeForString):
(WebKit::extractUnderlines):
(WebKit::compositionHighlights):
(WebKit::extractMarkedTextComposition):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setMarkedText:selectedRange:]):
(shouldExtractMarkedTextComposition):
(-[WKContentView setAttributedMarkedText:selectedRange:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::shouldExtractMarkedTextComposition):
(WebKit::WebViewImpl::setMarkedText):
(WebKit::extractInitialUnderlines): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
